File: SDK/CentralService/java/client/src/main/java/erp/centralservice/client/CentralServiceSdkOptions.java
package erp.centralservice.client;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

/**
 * Central Service 发现侧 SDK 的基础配置。
 *
 * <p>配置同时支持单 {@code baseUrl} 构造与多端点 {@code endpoints} 配置。
 * discovery 侧会按优先级顺序请求端点，并在发生传输层异常时按单端点最大尝试次数与熔断状态切换备用。</p>
 */
public final class CentralServiceSdkOptions {
    private static final int DEFAULT_MAX_ATTEMPTS = 2;

    /** 首个中心服务根地址（与 {@link #endpoints} 的首项等价）。 */
    public final String baseUrl;
    /** 按优先级归一化后的中心服务端点列表。 */
    public final List<CentralServiceEndpointOptions> endpoints;
    /** 请求连接与读取超时时间，单位毫秒；小于等于 0 时由 HTTP 层回退为默认值。 */
    public int timeoutMs = 5000;
    /** 发送给服务端的 User-Agent；为空白时不会写入请求头。 */
    public String userAgent = "centralservice-java-client/0.1.0";

    /**
     * 使用中心服务根地址创建配置对象。
     *
     * @param baseUrl 中心服务根地址，例如 {@code http://127.0.0.1:5000}
     */
    public CentralServiceSdkOptions(String baseUrl) {
        this(Collections.singletonList(new CentralServiceEndpointOptions(baseUrl)));
    }

    /**
     * 使用中心服务端点列表创建配置对象。
     *
     * @param endpoints 中心服务端点列表
     */
    public CentralServiceSdkOptions(List<CentralServiceEndpointOptions> endpoints) {
        if (endpoints == null) throw new IllegalArgumentException("endpoints is required");
        this.endpoints = normalizeEndpoints(endpoints);
        if (this.endpoints.isEmpty()) throw new IllegalArgumentException("at least one endpoint is required");
        this.baseUrl = this.endpoints.get(0).baseUrl;
    }

    private static List<CentralServiceEndpointOptions> normalizeEndpoints(List<CentralServiceEndpointOptions> endpoints) {
        ArrayList<CentralServiceEndpointOptions> normalized = new ArrayList<CentralServiceEndpointOptions>();
        for (int i = 0; i < endpoints.size(); i++) {
            CentralServiceEndpointOptions endpoint = endpoints.get(i);
            if (endpoint == null || endpoint.baseUrl == null || endpoint.baseUrl.trim().isEmpty()) {
                continue;
            }
            normalized.add(endpoint.normalize(i));
        }
        Collections.sort(normalized, new Comparator<CentralServiceEndpointOptions>() {
            @Override
            public int compare(CentralServiceEndpointOptions left, CentralServiceEndpointOptions right) {
                if (left.priority != right.priority) {
                    return left.priority < right.priority ? -1 : 1;
                }
                if (left.order == right.order) {
                    return 0;
                }
                return left.order < right.order ? -1 : 1;
            }
        });
        return Collections.unmodifiableList(normalized);
    }

    static String normalizeBaseUrl(String baseUrl) {
        if (baseUrl == null || baseUrl.trim().isEmpty()) throw new IllegalArgumentException("baseUrl is required");
        String s = baseUrl.trim();
        while (s.endsWith("/")) {
            s = s.substring(0, s.length() - 1);
        }
        return s;
    }

    public static int normalizeMaxAttempts(Integer maxAttempts) {
        int value = maxAttempts == null ? DEFAULT_MAX_ATTEMPTS : maxAttempts.intValue();
        return value < 1 ? DEFAULT_MAX_ATTEMPTS : value;
    }

    /**
     * 表示单个中心服务端点的连接与熔断配置。
     */
    public static final class CentralServiceEndpointOptions {
        /** 中心服务根地址。 */
        public final String baseUrl;
        /** 优先级，数值越小越优先。 */
        public int priority;
        /** 单中心最大尝试次数；为空时使用默认值 2。 */
        public Integer maxAttempts;
        /** 单中心熔断器配置；为空表示禁用熔断。 */
        public CentralServiceCircuitBreakerOptions circuitBreaker;
        int order;

        /**
         * 使用中心服务根地址创建端点配置。
         *
         * @param baseUrl 中心服务根地址
         */
        public CentralServiceEndpointOptions(String baseUrl) {
            this.baseUrl = baseUrl == null ? "" : baseUrl;
        }

        CentralServiceEndpointOptions normalize(int endpointOrder) {
            CentralServiceEndpointOptions normalized = new CentralServiceEndpointOptions(normalizeBaseUrl(baseUrl));
            normalized.priority = priority;
            normalized.maxAttempts = Integer.valueOf(normalizeMaxAttempts(maxAttempts));
            normalized.circuitBreaker = circuitBreaker == null ? null : circuitBreaker.normalize();
            normalized.order = endpointOrder;
            return normalized;
        }
    }

    /**
     * 表示单个中心服务端点的熔断配置。
     */
    public static final class CentralServiceCircuitBreakerOptions {
        /** 连续失败阈值。 */
        public int failureThreshold;
        /** 熔断持续时间，单位分钟。 */
        public int breakDurationMinutes;
        /** 半开状态恢复所需的连续成功阈值。 */
        public int recoveryThreshold;

        CentralServiceCircuitBreakerOptions normalize() {
            CentralServiceCircuitBreakerOptions normalized = new CentralServiceCircuitBreakerOptions();
            normalized.failureThreshold = Math.max(1, failureThreshold);
            normalized.breakDurationMinutes = Math.max(1, breakDurationMinutes);
            normalized.recoveryThreshold = Math.max(1, recoveryThreshold);
            return normalized;
        }
    }
}

File: SDK/CentralService/dotnet/src/CentralService.Service/CentralServiceSdkOptions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using CentralService.Shared.Internal;

namespace CentralService.Service
{
    /// <summary>
    /// 定义中心服务服务端 SDK 的连接选项。
    /// </summary>
    public sealed class CentralServiceSdkOptions
    {
        private const int DefaultMaxAttempts = 2;
        private readonly IReadOnlyList<CentralServiceEndpointOptions> _endpoints;

        /// <summary>
        /// 使用单个中心服务根地址初始化客户端选项。
        /// </summary>
        /// <param name="baseUrl">中心服务的根地址。</param>
        public CentralServiceSdkOptions(string baseUrl)
            : this(new[] { new CentralServiceEndpointOptions(baseUrl) })
        {
        }

        /// <summary>
        /// 使用多个中心服务端点初始化客户端选项。
        /// </summary>
        /// <param name="endpoints">中心服务端点列表。</param>
        public CentralServiceSdkOptions(IEnumerable<CentralServiceEndpointOptions> endpoints)
        {
            if (endpoints == null) throw new ArgumentNullException("endpoints");

            var normalized = NormalizeEndpoints(endpoints).ToArray();
            if (normalized.Length == 0)
            {
                throw new ArgumentException("至少需要一个中心服务端点。", "endpoints");
            }

            _endpoints = normalized;
            BaseUrl = _endpoints[0].BaseUrl;
            Timeout = TimeSpan.FromSeconds(5);
            IgnoreSslErrors = false;
        }

        /// <summary>
        /// 获取首个中心服务根地址（与 <see cref="Endpoints"/> 的首项等价）。
        /// </summary>
        public string BaseUrl { get; private set; }

        /// <summary>
        /// 获取按优先级归一化后的中心服务端点列表。
        /// </summary>
        public IReadOnlyList<CentralServiceEndpointOptions> Endpoints => _endpoints;

        /// <summary>
        /// 获取或设置单次请求超时时间。
        /// </summary>
        public TimeSpan Timeout { get; set; }

        /// <summary>
        /// 获取或设置是否忽略 SSL 证书错误。
        /// </summary>
        public bool IgnoreSslErrors { get; set; }

        /// <summary>
        /// 获取或设置自定义的 HTTP 消息处理器，用于在测试环境中与 TestServer 集成。
        /// </summary>
        public System.Net.Http.HttpMessageHandler? HttpMessageHandler { get; set; }

        internal IReadOnlyList<CentralServiceTransportEndpoint> CreateTransportEndpoints()
        {
            return _endpoints
                .Select((endpoint, index) => endpoint.ToTransportEndpoint(index))
                .ToArray();
        }

        private static IEnumerable<CentralServiceEndpointOptions> NormalizeEndpoints(IEnumerable<CentralServiceEndpointOptions> endpoints)
        {
            return endpoints
                .Where(x => x != null && !string.IsNullOrWhiteSpace(x.BaseUrl))
                .Select((x, index) => x.Normalize(index))
                .OrderBy(x => x.Priority)
                .ThenBy(x => x.Order)
                .ToArray();
        }

        /// <summary>
        /// 表示单个中心服务端点的连接与熔断配置。
        /// </summary>
        public sealed class CentralServiceEndpointOptions
        {
            /// <summary>
            /// 初始化一个空的端点配置实例，供配置系统反序列化使用。
            /// </summary>
            public CentralServiceEndpointOptions()
            {
                BaseUrl = string.Empty;
            }

            /// <summary>
            /// 使用中心服务根地址创建端点配置。
            /// </summary>
            public CentralServiceEndpointOptions(string baseUrl)
            {
                BaseUrl = baseUrl ?? string.Empty;
            }

            /// <summary>
            /// 获取或设置中心服务根地址。
            /// </summary>
            public string BaseUrl { get; set; }

            /// <summary>
            /// 获取或设置优先级，数值越小越优先。
            /// </summary>
            public int Priority { get; set; }

            /// <summary>
            /// 获取或设置单中心最大尝试次数；为空时使用系统默认值 2。
            /// </summary>
            public int? MaxAttempts { get; set; }

            /// <summary>
            /// 获取或设置该端点的熔断配置；为空时表示禁用熔断。
            /// </summary>
            public CentralServiceCircuitBreakerOptions? CircuitBreaker { get; set; }

            internal int Order { get; private set; }

            internal CentralServiceEndpointOptions Normalize(int order)
            {
                return new CentralServiceEndpointOptions(NormalizeBaseUrl(BaseUrl))
                {
                    Priority = Priority,
                    MaxAttempts = NormalizeMaxAttempts(MaxAttempts),
                    CircuitBreaker = CircuitBreaker?.Normalize(),
                    Order = order,
                };
            }

            internal CentralServiceTransportEndpoint ToTransportEndpoint(int order)
            {
                var maxAttempts = NormalizeMaxAttempts(MaxAttempts);
                var circuitBreakerState = CircuitBreaker == null
                    ? null
                    : new CentralServiceCircuitBreakerState(
                        CircuitBreaker.FailureThreshold,
                        TimeSpan.FromMinutes(CircuitBreaker.BreakDurationMinutes),
                        CircuitBreaker.RecoveryThreshold);

                return new CentralServiceTransportEndpoint(NormalizeBaseUrl(BaseUrl), Priority, maxAttempts, order, circuitBreakerState);
            }

            private static string NormalizeBaseUrl(string baseUrl)
            {
                if (string.IsNullOrWhiteSpace(baseUrl))
                {
                    throw new ArgumentNullException("baseUrl");
                }

                return baseUrl.Trim().TrimEnd('/');
            }

            private static int NormalizeMaxAttempts(int? maxAttempts)
            {
                var value = maxAttempts.GetValueOrDefault(DefaultMaxAttempts);
                return value < 1 ? DefaultMaxAttempts : value;
            }
        }

        /// <summary>
        /// 表示单个中心服务端点的熔断配置。
        /// </summary>
        public sealed class CentralServiceCircuitBreakerOptions
        {
            /// <summary>
            /// 获取或设置连续失败阈值。
            /// </summary>
            public int FailureThreshold { get; set; }

            /// <summary>
            /// 获取或设置熔断持续时间（分钟）。
            /// </summary>
            public int BreakDurationMinutes { get; set; }

            /// <summary>
            /// 获取或设置半开状态下的恢复成功阈值。
            /// </summary>
            public int RecoveryThreshold { get; set; }

            internal CentralServiceCircuitBreakerOptions Normalize()
            {
                return new CentralServiceCircuitBreakerOptions
                {
                    FailureThreshold = Math.Max(1, FailureThreshold),
                    BreakDurationMinutes = Math.Max(1, BreakDurationMinutes),
                    RecoveryThreshold = Math.Max(1, RecoveryThreshold),
                };
            }
        }
    }
}
